### ISSUE TITLE
parsing support for unquoted reserved keyword identifiers

### ISSUE THREAD

**Issue by krahnikblis** [sqlfluff/sqlfluff#4817]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

fix unparsable SQL enough to make it parsable, or adjust the parser to allow unquoted reserved words to be used as identifiers (in line with database engines allowing for the same).

### Use case

i'm plugging existing operational ETL SQLs into SQLFluff to prettify and standardize them and also build a lineage graph, but in some cases SQLFluff cannot parse the statements due to having more restrictive rules than the db engines that successfully handle these scripts. one common scenario is when SQL keywords are used as column names, for example `order` (as in a sales order number). often the authors of these SQLs don't abide by [a whole bunch of] formatting best practices, and don't quote reserved words when used as identifiers. they don't run into problems in the database because no error is thrown; i.e. this is operational production SQL. my expectation when parsing and fixing their scripts is that SQLFluff could fix these things when it finds them, but instead it just fails to parse the statements altogether.

here's my current workaround, perhaps something like this could be put into the parser, to recursively try to fix PRS errors by checking each time whether the part of the string it stumbles on contains an unquoted reserved word used as an identifier. i'm sure there are holes in my approach, and it might break something else, but i'm using it in a way that it just is attempting to recover from an otherwise lost cause:

```python
from sqlfluff.core.dialects import dialect_selector

def repair_reserved_unparsable(sql, dialect, lint):
    # get reserved words for this dialect
    reserved = [w.lower() for w in dialect_selector(s=dialect).sets("reserved_keywords")]
    replacements = []
    for d in lint:
        # pull the stumbling string out of the lint PRS-coded error description
        unparsable = re.search("(?<=Found unparsable section: ')(.*)(?=\.\.\.')",d["description"])
        # fix encoding because any tabs/newlines/etc will have double-escape
        unparsable = codecs.escape_decode(unparsable[0], 'hex')[0].decode() 
        parsable = unparsable
        # check each word in the stumbling string against the list of keywords - any match means it's both unquoted and a reserved word thus a likely cause of the parse error. if found, replace with quoted self, and put before/after stumble string in a list
        for w in re.split(" |(?<=\.)|,|\s",unparsable):
            if w != "" and w in reserved:
                parsable = parsable.replace(w,"`" + w + "`")
        replacements.append([unparsable,parsable])
    newsql = sql
    # replace the whole stumble string (which has the replaced keyword) to minimize likelihood of replacing the wrong instance of the keyword in the large SQL statement
    for r in replacements:
        newsql = newsql.replace(r[0], r[1])
    if len(r) > 0:
        print("attempted fix of unquoted reserved keywords to prevent parsing error in fix")
    else:
        print("parsing errors exist in lint but not caused by unquoted dialect reserved words")
    return newsql
```
and later on in a function:
```python
sql_orig = node["sqlText"]
dialect = node["job_type"] if node["job_type"] in cfg["sql_dialects"] else cfg["sql_dialect_default"]
# lint first, to fix parsing
linted = sqlfluff.lint(sql_orig, dialect=dialect, config_path=cfg["sqlfluff_cfg"]) 
# any parse errors? put them in a list to try to fix
prs_lint = [l for l in linted if l["code"] == "PRS"] # focus on PRS errors that might be due to keyword
if len(prs_lint) > 0:
  # use the keyword quoting function to try to fix the statement
  sqlText = repair_reserved_unparsable(sql_orig, dialect, prs_lint)
else:
  sqlText = sql_orig
try: # now when running a full `fix`, it won't fail due to parse error:
                sql_fmt = sqlfluff.fix(sqlText, dialect=dialect, config_path=cfg["sqlfluff_cfg"])
```

### Dialect

all

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by WittierDinosaur** [comment]
Hey @krahnikblis, can you please provide specific examples of reserved keywords in dialects are parsable. The difference between a reserved and an unreserved keyword is that a reserved keyword cannot be used as an identifier, and an unreserved can. What you're probably finding here is keywords that we've listed as reserved for a dialect, where they're actually not

**Comment by krahnikblis** [comment]
@WittierDinosaur sure, i'm seeing this mostly in Spark scripts, but that's probably because most of the jobs in our warehouse are Spark (Hive is the next most common). if it would help, i could write the replacement keywords to a log when i run that above function against all my prior failed parsings, (i've got 98 SQL files dumped due to some error or t'other, not all are caused by this but it should be a good sample), and share the ones that end up fixing the problem for parsing the scripts... 

here's a simple sample showing how `order` as a column name chokes it up:
```python
sparksql = """insert into some_schema.target_table
select distinct a.rubber, a.ducky, a.sales, a.order, a.bathtub
from some_schema.source_table as a;"""
linted = sqlfluff.lint(sparksql, dialect="sparksql", config_path=cfg["sqlfluff_cfg"]) 
linted
```
produces this:
```
 {'line_no': 2,
  'line_pos': 46,
  'code': 'PRS',
  'description': "Line 2, Position 46: Found unparsable section: '.order, a.bathtub'",
  'name': ''},
```

tried with a few of the more common ones. Hive and Teradata both had the same parsing error (but i don't remember if the actual DBs would allow `order` as a column name), Postgres worked fine

**Comment by peterbud** [comment]
Hi @krahnikblis,

Thanks for raising this! It is definitely a common and frustrating pain point when dealing with legacy, operational production SQL. It's completely understandable to expect a linter/parser to gracefully handle code that the database engine itself executes without a hitch.

To help clarify why SQLFluff behaves this way—and how you can work around it—it helps to look under the hood at how the parser interacts with different SQL dialects, alongside the general philosophy of linting.

**1. Dialects, Reserved vs. Unreserved Keywords, and Quoting**
SQLFluff is highly dependent on the specific dialect you configure (e.g., snowflake, bigquery, postgres). Each database engine has its own complex implementation of what constitutes a reserved keyword vs. an unreserved keyword.

* Reserved Keywords: Database engines strictly forbid these as identifiers unless they are explicitly quoted (e.g., "order", `order`).
* Unreserved Keywords: These can be used as identifiers without quotes in certain contexts, depending on the engine's internal grammar state machine.

Because SQLFluff aims to mimic the database's actual parser, if a word like ORDER is strictly reserved in that dialect and appears unquoted as a column name, the parser physically cannot construct an Abstract Syntax Tree (AST). If it can't parse the file, it cannot apply fixing rules. A linter cannot fix what it cannot first understand.

**2. The Philosophy: Minimal Intervention & Augmentation**
When it comes to formatting and auto-fixing, SQLFluff operates on two core principles:
* **Minimal Intervention**: The primary goal of a linter is to ensure code meets a specific standard (rules), not necessarily to rewrite it into a singular "canonical" form. If your code satisfies the configured rules, the linter views the existing version as a "legal" expression of that code.
* **Augmentation over Replacement**: Think of the linter as a set of guardrails. It prevents you from drifting into "bad" territory (like lines being too long or inconsistent casing), but it still leaves the "driving" (the specific aesthetic arrangement and syntax choices, like choosing to quote or not quote) to the developer.

Expecting the linter to automatically intercept a fatal parsing error and rewrite the underlying syntax structure blurs the line between a linter and a transpiler.

**3. Available Solutions & Workarounds**
If the database engine genuinely allows order as an unquoted identifier in your specific context, it implies SQLFluff's dialect definition might be slightly too restrictive compared to the real-world engine.

While the team evaluates dialect adjustments, you can unblock yourself immediately using SQLFluff’s configuration:

Refuse Token / Keyword Rules (RF04)
If the parsing failure is being triggered by rule RF04 (which governs the use of unquoted identifiers or specific reserved words), you can explicitly tell SQLFluff to ignore certain keywords in your .sqlfluff configuration file using the ignore_words parameter:

```ini
[sqlfluff:rules:references.keywords] # RF04
ignore_words = order, group, user
```
